maldoca/ole/oss_utils.cc

```diff
@@ -438,6 +436,67 @@ bool BufferToUtf8::ConvertEncodingBufferToUTF8String(absl::string_view input,
     return true;
   }
   const char* input_ptr = input.data();
+
+#if defined(_WIN32)
+  int mb_size = 0; // size of the multi-byte string
+  int wc_size = 0; // size of the wide-character string
+  std::unique_ptr<wchar_t*> wc_str;
+  bool is_already_utf16 = code_page_ == kUtf16LECodePage;
+
+  // No need to convert to UTF-16, if the string is already UTF-16 encoded.
+  if (!is_already_utf16) {
+    // Calculate output size in UTF-16.
+    wc_size =
+        MultiByteToWideChar(code_page_, 0, input_ptr, input.size(), NULL, 0);
+    DLOG(INFO) << "Output size in UTF-16: " << wc_size;
+    if (wc_size <= 0) {
+      LOG(ERROR) << "Error while calculating output size in UTF-16: "
+                 << GetLastError();
+      return false;
+    }
+
+    // Allocate memory for the temp UTF-16 output and convert input to UTF-16.
+    wc_str = std::make_unique<wchar_t*>(new wchar_t[wc_size]);
+    wc_size = MultiByteToWideChar(code_page_, 0, input_ptr, input.size(), *wc_str.get(),
+                                  wc_size);
+    if (wc_size <= 0) {
+      LOG(ERROR) << "Error while converting to UTF-16: " << GetLastError();
+      return false;
+    }
+  } else {
+    wc_str = std::make_unique<wchar_t*>((wchar_t*)input.data());
+    wc_size = input.size() / 2;
+  }
+
+  // Calculate output size in UTF-8.
+  mb_size =
+      WideCharToMultiByte(CP_UTF8, 0, *wc_str.get(), wc_size, NULL, 0, NULL, NULL);
+  DLOG(INFO) << "Output size in UTF-8: " << mb_size;
+  if (mb_size <= 0) {
+    LOG(ERROR) << "Error while calculating output size in UTF-8: "
+               << GetLastError();
+    return false;
+  }
+
+  // Allocate proper memory and convert input to UTF-8.
+  out_str->resize(mb_size);
+  mb_size = WideCharToMultiByte(CP_UTF8, 0, *wc_str.get(), wc_size, &(*out_str)[0],
+                                mb_size, NULL, NULL);
+  if (mb_size <= 0) {
+    LOG(ERROR) << "Error while converting to UTF-8: " << GetLastError();
+    return false;
+  }
+
+  // TODO: fallback to internal converters if available.
+
+  // For some reason, it preserves start and trailing \0 so remove them
+  StripNullChar(out_str);
```

Comment:
That doesn't sound right.  I can see how it may get a 0 at the end but not in the front.  Please check again and if you get a 0 at the end, just resize the string back.

Reply:
the comment is copied from the linux version: https://github.com/google/maldoca/blob/master/maldoca/ole/oss_utils.cc#L502

not sure if resize is a better option as the function removes between 0 and multipled \0 depending on the input.

the behaviors is the same for win + linux. hence, i dont think it is related to the conversion apis but something with the parser or the actual office files.

Reply:
ok.  it it's a parser issue . I have no idea why.